ExitIntegrationTest.java

```diff
@@ -0,0 +1,176 @@
+    
+    IGate testGate;
+    CarSensor testIS;
+    CarSensor testOS; 
+    ExitUI testUI;
+    Carpark testCarpark;
+    AdhocTicket  testAdhoc;
+    SeasonTicket testSeason;
+    IAdhocTicketDAO mockAdhocDAO;
+    ISeasonTicketDAO mockSeasonDAO;
+    ExitController testExit;
+    String seasonTicketId, cParkName;
+    UsageRecord mockRecord;
+    int cParkCap;
+    
+    @Before
+    public void setUp() {
+        cParkName = "test name";
+        cParkCap = 5;
+        mockAdhocDAO = mock(AdhocTicketDAO.class);
+        mockSeasonDAO = mock(SeasonTicketDAO.class);
+        mockRecord = mock(UsageRecord.class);
+        testCarpark = new Carpark(cParkName, cParkCap, mockAdhocDAO, mockSeasonDAO);
+        testGate = new Gate(0, 0);
+        testIS = new CarSensor("test inside", 100, 100);
+        testOS = new CarSensor("test outside", 200, 200);
+        testUI = mock(ExitUI.class);
+        testAdhoc = new AdhocTicket(cParkName, 1, "Atest");
+        testSeason = new SeasonTicket("test", cParkName, 1234, 4321);
+        testExit = new ExitController(testCarpark, testGate, testIS, testOS, testUI);
+    }
+    
+    @Test
+    public void initialisationTest() {
+        assertTrue(testCarpark instanceof Carpark);
+        assertTrue(testIS instanceof CarSensor);
+        assertTrue(testOS instanceof CarSensor);
+        assertTrue(testExit instanceof ExitController);
+    }
+    
+    @Test
+    public void testNormalFlowSeason() {
+        assertEquals(testExit.getState(), "IDLE");  //checks initial state
+        testIS.setCarDetected(true);
+        testExit.carEventDetected(testIS.getId(), true);
+        
+        assertEquals(testExit.getState(), "WAITING");
+        verify(testUI).display("Insert Ticket");
+        
+        when(mockSeasonDAO.findTicketById(testSeason.getId())).thenReturn(testSeason);
+        testSeason.recordUsage(mockRecord);
+        testExit.ticketInserted(testSeason.getId());
+        assertEquals(testExit.getState(), "PROCESSED");
+        
+        testExit.ticketTaken();
+        verify(testUI).display("Take Processed Ticket");
+        assertEquals(testExit.getState(), "TAKEN");     
+        
+        testOS.setCarDetected(true);
+        testExit.carEventDetected(testOS.getId(), true);
+        assertEquals(testExit.getState(), "EXITING");
+        
+        testIS.setCarDetected(false);
+        testExit.carEventDetected(testIS.getId(), false);
+        assertEquals(testExit.getState(), "EXITED");
+    }
+    
+    @Test
+    public void testNormalFlowAdhoc() {
+        assertEquals(testExit.getState(), "IDLE");  //checks initial state
+        testIS.setCarDetected(true);
+        testExit.carEventDetected(testIS.getId(), true);
+        
+        assertEquals(testExit.getState(), "WAITING");
+        verify(testUI).display("Insert Ticket");            //checks state transition and message display
+        
+        when(mockAdhocDAO.findTicketByBarcode("Atest")).thenReturn(testAdhoc);

```

Comment:
This line is wrong.
this is a test of the system, so you need to prove the system is doing what it is supposed it. So you cannot use a mock and say whenXthendoY you have to use verify or assert and show that it is actually doing what mocked in the unit tests
In this case I think it should be assertEquals(mockAdhocDAO.findTicketByBarcode("Atest"), testAdhoc);

Reply:
In this case, I felt it was appropriate, since the AdhocTicketDAO and such was part of Aaron's testing, mine mainly focused on ExitController, it's communication with Carpark, and SeasonTicket. I didn't strictly adhere to that unfortunately, but this is the reason for using a mock here.

Reply:
fair enough, maybe put a comment saying what your doing?